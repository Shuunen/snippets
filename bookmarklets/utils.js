/* eslint-disable-next-line no-unused-vars */
class Shuutils {
  constructor (app) {
    app = app || {
      id: 'shu-app',
      debug: true
    }
    this.app = app
    this.accentsIn = 'ÀÁÂÃÄÅĄàáâãäåąßÒÓÔÕÕÖØÓòóôõöøóÈÉÊËĘèéêëęðÇĆçćÐÌÍÎÏìíîïÙÚÛÜùúûüÑŃñńŠŚšśŸÿýŽŻŹžżź'
    this.accentsOut = 'AAAAAAAaaaaaaaBOOOOOOOOoooooooEEEEEeeeeeeCCccDIIIIiiiiUUUUuuuuNNnnSSssYyyZZZzzz'
  }

  log (...stuff) {
    if (!this.app.debug) {
      return
    }
    stuff.unshift(this.app.id + ' :')
    console.log.apply(console, stuff)
  }

  warn (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.warn.apply(console, stuff)
  }

  error (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.error.apply(console, stuff)
  }

  readableString (str) {
    return str.split('').map(function (letter) {
      var i = this.accentsIn.indexOf(letter)
      return i !== -1 ? this.accentsOut[i] : letter
    }).join('').replace(/(\W|-)/gi, ' ').replace(/\s+/, ' ')
  }

  debounce (func, wait, immediate) {
    var timeout
    return function () {
      var context = this
      var args = arguments
      var later = function later () {
        timeout = null
        if (!immediate) {
          func.apply(context, args)
        }
      }
      var callNow = immediate && !timeout
      clearTimeout(timeout)
      timeout = setTimeout(later, wait)
      if (callNow) {
        func.apply(context, args)
      }
    }
  }

  findOne (selector, context, dontYell) {
    context = context || document
    var item = context.querySelector(selector)
    if (item && this.app.debug) {
      this.log('found element matching "' + selector + '"')
    } else if (!item && !dontYell) {
      this.warn('found no element for selector "' + selector + '"')
    }
    return item
  }

  findFirst (selector, context, dontYell) {
    return this.findAll(selector, context, dontYell)[0]
  }

  findAll (selector, context, dontYell) {
    if (!selector || !selector.length || selector.length === 1) {
      this.error('incorrect selector : ', selector)
    }
    context = context || document
    var items = Array.prototype.slice.call(context.querySelectorAll(selector))
    if (items.length && this.app.debug) {
      this.log('found', items.length, 'elements matching "' + selector + '"')
    } else if (!items.length && !dontYell) {
      this.warn('found no elements for selector "' + selector + '"')
    }
    return items
  }
}
