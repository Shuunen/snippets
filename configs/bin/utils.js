import { copyFile, mkdir } from 'node:fs/promises'
import path from 'node:path'

/**
 * Get the filename from a filepath
 * @param {string} filepath the filepath to get the filename from
 * @returns {string} the filename
 * @example filename('C:\\Users\\me\\file.txt') // 'file.txt'
 */
export function filename(filepath = '') {
  return /[/\\](?<name>[\w.-]+)$/u.exec(filepath)?.groups?.name ?? ''
}

/**
 * Remove lines matching a regexp list
 * @param {string} content the content to clean
 * @param {RegExp[]} regexList the regex list to match
 * @returns {string} the cleaned content
 */
export function removeLinesMatching(content, regexList) {
  const lines = content.split('\n')
  const filteredLines = lines.filter(line => !regexList.some(regex => regex.test(line)))
  return filteredLines.join('\n').trim()
}

/**
 * Remove lines after a given regex
 * @param {string} content the content to remove lines from
 * @param {RegExp} regex the regexp to match the line after which to remove content
 * @returns {string} the content without the lines after the matching line
 */
export function removeLinesAfter(content, regex) {
  const lines = content.split('\n')
  const index = lines.findIndex(line => regex.test(line))
  if (index === -1) return content.trim()
  return lines.slice(0, index).join('\n').trim()
}

/**
 * Convert carriage return to unix line endings
 * @param {string} content the content to be processed
 * @returns the processed content with unix line endings
 */
export function useUnixCarriageReturn(content) {
  return content.replace(/\r\n/gu, '\n')
}

/**
 * Clean a file details content
 * @param {string} content the file content to clean
 * @param {RegExp} [linesAfter] a regex to remove lines after
 * @param {RegExp[]} [linesMatching] a list of regex to remove lines matching
 * @param shouldClearSpaces if true will also clear spaces
 * @returns {string} the cleaned file content
 */
// eslint-disable-next-line @typescript-eslint/max-params
export function clean(content, linesAfter, linesMatching, shouldClearSpaces = true) {
  if (!content) return ''
  let output = content
  if (linesAfter) output = removeLinesAfter(output, linesAfter)
  if (linesMatching) output = removeLinesMatching(output, linesMatching)
  if (shouldClearSpaces) output = output.replace(/\s*/gu, '')
  return output
}

/* c8 ignore start */

/**
 * Normalize a filepath with slash style
 * @param {string} filepath the filepath
 * @param {boolean} shouldUseTilde use tilde will replace the home directory with ~
 * @param {string} home the home directory path
 * @returns the normalized path
 */
export function normalizePathWithSlash(filepath, shouldUseTilde = false, home = process.env.HOME ?? '') {
  let outPath = path.normalize(filepath).replace(/\\/gu, '/')
  if (shouldUseTilde) outPath = outPath.replace(normalizePathWithSlash(home), '~')
  return outPath
}

/**
 * Copy a file
 * @param {string} source the source file
 * @param {string} destination the destination file
 * @returns {Promise<boolean>} some bool result; i dont know im in the train to Paris
 */
export async function copy(source, destination) {
  // destination will be created or overwritten by default.
  const destinationFolder = destination.replace(filename(destination), '')
  // eslint-disable-next-line @typescript-eslint/naming-convention
  await mkdir(destinationFolder, { recursive: true })

  return copyFile(source, destination)
    .then(() => true)
    .catch(error => {
      console.log(error)
      return false
    })
}
/* c8 ignore stop */
