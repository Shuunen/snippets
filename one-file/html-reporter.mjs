import { gray, green } from 'shuutils/dist/colors'

const states = {
  lookingForATag: 'lookingForATag',
  onTagName: 'onTagName',
  onTagAttr: 'onTagAttr',
  scanComplete: 'scanComplete',
}

export class HtmlReporter {
  constructor (input = '', debug = false) {
    this.input = input
    this.index = -1
    this.debug = debug
    this.attr = 0
    this.css = 0
    this.styles = 0
    this.tags = 0
    this.text = 0
    this.total = input.length
    this.state = ''
    this.setState(states.lookingForATag)
    this.scan()
  }
  scan () {
    /* eslint-disable curly */
    this.index++
    const char = this.input[this.index]
    if (!char || this.index === this.total) return this.onScanComplete()
    if (char === '>') {
      this.tags++
      this.setState(states.lookingForATag)
    } else if (this.state === states.onTagName && char === ' ') {
      this.attr++
      this.setState(states.onTagAttr)
    } else if (this.state === states.onTagName && char === 's') {
      const [match] = this.input.slice(this.index).match(/^style[\S\s]*?<\/style>/) || []
      if (match) {
        this.tags += 14 // the "<" has already been count on stats.tags, it remains "style></style>" to be count as stats.tags
        this.css += match.length - 14 // here stats.css is only the content
        this.index += match.length - 1
      } else this.tags++
    } else if (this.state === states.onTagName) {
      this.tags++
    } else if (this.state === states.onTagAttr && char === 's') {
      const [match] = this.input.slice(this.index).match(/^style="[^"]+"/) || []
      if (match) {
        this.styles += match.length
        this.index += match.length - 1
      } else this.attr++
    } else if (this.state === states.onTagAttr && char !== '>') {
      this.attr++
    } else if (this.state === states.lookingForATag && char !== '<') {
      this.text++
    } else if (char === '<') {
      this.tags++
      this.setState(states.onTagName)
    }
    /* eslint-enable curly */
    this.scan()
  }
  onScanComplete () {
    this.setState(states.scanComplete)
    console.assert((this.tags + this.attr + this.text + this.styles + this.css) === this.total, 'sub-stats does not adds up to the total')
  }
  /**
   * @param {number} index
   * @param {(char: string) => string} color
   * @returns
   */
  readable (index, color) {
    let char = this.input[index]
    if (!char) return ' '
    if (char === '\n') char = '\\n'
    if (char === ' ') char = '\\s'
    return color(char)
  }
  /**
   * Set the state of the scanner
   * @param {string} newState
   */
  setState (newState) {
    if (this.debug) {
      const context = this.readable(this.index - 2, gray) + this.readable(this.index - 1, gray) + this.readable(this.index, green) + this.readable(this.index + 1, gray) + this.readable(this.index + 2, gray)
      const stateChange = `${this.state} ${gray('=>')} ${newState}`
      const stats = `tags:${this.tags.toString().padEnd(3)} attr:${this.attr.toString().padEnd(3)} text:${this.text.toString().padEnd(3)}`
      console.log(`${this.index.toString().padEnd(4)} ${stateChange.padEnd(28)} ${stats.padEnd(24)} ${context}`)
    }
    this.state = newState
  }
}
