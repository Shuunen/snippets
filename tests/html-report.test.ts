import { gray } from 'shuutils'
import { HtmlReporter } from '../one-file/html-reporter.mjs'
import { check, checkSnapshot } from './utils.js'

const samples = [
  {
    input: '',
    output: { attr: 0, css: 0, styles: 0, tags: 0, text: 0, total: 0 },
    title: 'empty',
  },
  {
    input: '<br/><style>h1 { color: red; font-size: 2rem }</style>',
    output: { attr: 0, css: 34, styles: 0, tags: 20, text: 0, total: 54 },
    title: 'empty',
  },
  {
    input: '<h1 style="color: red">Super top !</h1>',
    output: { attr: 1, css: 0, styles: 18, tags: 9, text: 11, total: 39 },
    title: 'empty',
  },
  {
    input: '<p><small>ah</small></p> ',
    output: { attr: 0, css: 0, styles: 0, tags: 22, text: 3, total: 25 },
    title: 'empty',
  },
  {
    input: '<p style="font-weight: bold; color: black;"><small style="font-style: italic">ah</small></p> ',
    output: { attr: 2, css: 0, styles: 66, tags: 22, text: 3, total: 93 },
    title: 'empty',
  },
  {
    input: `<p>
<small>ah... !!</small>
</p>`,
    output: { attr: 0, css: 0, styles: 0, tags: 22, text: 10, total: 32 },
    title: 'empty',
  },
  {
    input: `<style data-vue-ssr-id="d706d280:0 2a082f94:0">
*,
::after,
::before {
  box-sizing: border-box;
}
</style>
<link rel="preload" href="/_nuxt/static/1626092993/state.js" as="script">`,
    output: { attr: 67, css: 92, styles: 0, tags: 22, text: 0, total: 181 },
    title: 'empty',
  },
] as const

samples.forEach(({ input, output, title }) => {
  const actual = new HtmlReporter(input)
  check(`html-report ${title} attr`, actual.attr, output.attr)
  check(`html-report ${title} css`, actual.css, output.css)
  check(`html-report ${title} styles`, actual.styles, output.styles)
  check(`html-report ${title} tags`, actual.tags, output.tags)
  check(`html-report ${title} text`, actual.text, output.text)
  check(`html-report ${title} total`, actual.total, output.total)
})

const { input, output, title } = samples[5]
const actualDebug = new HtmlReporter(input, true)
check(`html-report debug ${title} attr`, actualDebug.attr, output.attr)
check(`html-report debug ${title} css`, actualDebug.css, output.css)
check(`html-report debug ${title} styles`, actualDebug.styles, output.styles)
check(`html-report debug ${title} tags`, actualDebug.tags, output.tags)
check(`html-report debug ${title} text`, actualDebug.text, output.text)
check(`html-report debug ${title} total`, actualDebug.total, output.total)
checkSnapshot('html-report debug readable A', actualDebug.readable(3, gray))
checkSnapshot('html-report debug readable B', actualDebug.readable(16, gray))
