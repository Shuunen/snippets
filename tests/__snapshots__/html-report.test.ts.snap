// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html-report debug readable A 1`] = `"[90m\\\\n[39m"`;

exports[`html-report debug readable B 1`] = `"[90m\\\\s[39m"`;
